internal/npm_install/generate_build_file.js

```diff
@@ -0,0 +1,276 @@
+  pkgs.forEach(pkg => flattenDependencies(pkg, pkg, pkgs));
+
+  let buildFile = generatedHeader + allFilegroup;
+  pkgs.filter(pkg => !pkg._isNested).forEach(pkg => buildFile += printPackage(pkg));
+  fs.writeFileSync('BUILD.bazel', buildFile);
+}
+
+module.exports = {main};
+
+function isPackage(d) {
+  return fs.statSync(d).isDirectory() &&
+    fs.existsSync(path.join(d, 'package.json')) &&
+    fs.statSync(path.join(d, 'package.json')).isFile() &&
+    !fs.existsSync(path.join(d, 'WORKSPACE'));
+}
+
+function getPackagesIn(p) {
+  if (!fs.existsSync(p) || !fs.statSync(p).isDirectory()) {
+    return [];
+  }
+  return fs.readdirSync(p)
+    .map(f => path.join(p, f))
+    .filter(d => isPackage(d));
+}
+
+function findScopes(p) {
+  if (!fs.existsSync(p) || !fs.statSync(p).isDirectory()) {
+    return [];
+  }
+  return fs.readdirSync(p)
+    .filter(f => f.startsWith("@"))
+    .map(f => path.join(p, f))
+    .filter(d => fs.statSync(d).isDirectory());
+}
+
+function flattenDependencies(pkg, dep, pkgs) {
+  if (pkg._dependencies.indexOf(dep) !== -1) {
+    // circular dependency
+    return;
+  }
+  pkg._dependencies.push(dep);
+  const findDeps = function(targetDeps, required) {
+    Object.keys(targetDeps || {})
+        .map(targetDep => {
+          let matching = pkgs.filter(p => p._dir == `${dep._rootDir}/node_modules/${targetDep}`);
+          if (matching.length) {
+            // nested dep
+            return matching[0];
+          }
+          matching = pkgs.filter(p => p._dir == targetDep);
+          if (matching.length) {
+            return matching[0];
+          } else {
+            if (required) {
+              throw new Error(`Could not find required dep ${targetDep} of ${dep._dir}`)
+            }
+            return null;
+          }
+        })
+        .filter(dep => !!dep)
+        .map(dep => flattenDependencies(pkg, dep, pkgs));
+  };
+  findDeps(dep.dependencies, true);
+  findDeps(dep.peerDependencies, true);
+  findDeps(dep.optionalDependencies, false);
+}
+
+/**
+ * Reformat/pretty-print a json object as a skylark comment (each line
+ * starts with '# ').
+ */
+function printJson(pkg) {
+  // Clone and modify _dependencies to avoid circular issues when JSONifying
+  const cloned = {...pkg};
+  cloned._dependencies = cloned._dependencies.map(dep => dep._dir);
+  const result = JSON.stringify(cloned, null, 2).split('\n').map(line => `# ${line}`);
+  return result.join('\n');
+}
+
+/**
+ * Given a pkg, print a skylark `node_module` rule.
+ */
+function printPackage(pkg) {
+  const filegroup = `
+# Generated target for npm package "${pkg._dir}"
+${printJson(pkg)}
+filegroup(
+    name = "${pkg._dir}",
+    srcs = [
+        # ${pkg._dir} package contents (and contents of nested node_modules)
+        ":${pkg._dir}__files",
+        # direct or transitive dependencies hoisted to root by the package manager
+        ${
+      pkg._dependencies.filter(dep => dep != pkg)
```

Comment:
is dumping all transitive dependencies into the `srcs` attribute of a `filegroup` rule the canonical way of doing this? If we made a new `node_module` rule (or expose the `js_library` rule that exists in `internal`), when we create the runfiles object to return, there's a field called 'transitive_files' (https://docs.bazel.build/versions/master/skylark/lib/ctx.html#runfiles), which seems more canonical.

Is there a difference? Maybe just small perf gains?